Added TypeScript types and fixed minor bug

marcogrcr opened the pull request:
- Added a commit that updates it to version 1.0.2 published in https://www.npmjs.com/package/safeobj
- Added a commit to support TypeScript 2.2.0+
- Fixed bug when accessing defined `null`/`undefined` properties

marcogrcr committed:
updated to version 1.0.2

marcogrcr committed:
added typescript types

marcogrcr committed:
fixed bug on defined null/undefined properties

marcogrcr committed:
bumped up version to 1.0.3

gmmorris commented:
Hey,
I appreciate the PR but this repo is here as a Code sample for the corresponding blog post, so I'd rather not change it.
Also, it isn't clear why you removed the "Undefined" type which is quite integral to understanding the pattern described in the article... so I suspect your motivation for the PR isn't aligned with the motivation behind the original article, which is to describe the monadic approach to safe object access.

Feel free to fork the repo and publish a modified implementation if you want to.

marcogrcr commented:
Hi! Thanks for the response:

> Also, it isn't clear why you removed the "Undefined" type which is quite integral to understanding the pattern described in the article...

I'm not sure what you mean by removing the "Undefined" type.

I noticed that b9a0a32 added the `isUndefined()` function, yet [safeobj@1.0.1](https://www.npmjs.com/package/safeobj/v/1.0.1), didn't include it. Additionally another version, [safeobj@1.0.2](https://www.npmjs.com/package/safeobj/v/1.0.2), was also published, but with the exact script content as `1.0.1`, so I created 86111f0 to match what was on `npm`.

> I suspect your motivation for the PR isn't aligned with the motivation behind the original article, which is to describe the monadic approach to safe object access.

I added e7d9d68 to support TypeScript, since I wanted to consume your package, but my project is in TypeScript, which was the original motivation for this pull request. I believe this does not deviate from the original blog post, but rather it's a complement to reach a greater audience working in alternative languages.

> Feel free to fork the repo and publish a modified implementation if you want to.

I could do that yes, but I'd rather contribute to improve this package if possible, since [there are people out there using it](https://www.npmjs.com/package/safeobj).

The reason why I added 77f134d is because IMHO, not preventing *Uncaught TypeError: Cannot read property '🚀' of undefined* errors when the key **does exist**, yet it has `undefined`/`null` as it's value, defeats the purpose of this package which is to make nested object access safe. Without this change:

```js
let obj = {
  "name": "Luke",
  "mother": undefined
}

safe(obj).father.name === Undefined; // true
safe(obj).mother.name === Undefined; // throws Cannot read property 'name' of undefined
```

If you would be willing to accept the pull request, provided some modifications are made to it, let me know and I will gladly adjust it accordingly.

gmmorris commented:
Sorry for the early comments - I rushed through the PR and misread it.

LGTM and thanks for contributing!

aganoza commented:
hello thanks for the work, really liked this approach instead of lodash.get. 
Can this patch be published in npm as 1.0.3? Just for convenience.

gmmorris commented:
Ofcourse, sorry about the oversight.

Done
